lib/spack/spack/environment.py

```diff
@@ -1289,14 +1403,19 @@ def install_all(self, args=None):
         # a large amount of time due to repeatedly acquiring and releasing
         # locks, this does an initial check across all specs within a single
         # DB read transaction to reduce time spent in this case.
-        uninstalled_specs = []
+        specs_to_install = []
         with spack.store.db.read_transaction():
             for concretized_hash in self.concretized_order:
                 spec = self.specs_by_hash[concretized_hash]
-                if not spec.package.installed:
-                    uninstalled_specs.append(spec)
-
-        for spec in uninstalled_specs:
+                if not spec.package.installed or any(
+                    spec.satisfies(d) for d in self.dev_specs
+                ):
+                    specs_to_install.append(spec)
+
+        # Sort by total deps so specs have to be installed after dependencies
+        # This ensures dev-builds done in appropriate order
```

Comment:
This was for the case in which the dependency was itself a root spec of the environment, to ensure that `hdf5` would build before `trilinos`, for example. But I don't think it's necessary because of some improvements I've made, I should be able to revert that change.